presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/SetOperationFlattening.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.Iterables;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule.Context;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.SetOperationNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+abstract class SetOperationFlattening
```

Comment:
Instead of using inheritance for code reuse, it'd be cleaner to model this as a method that takes a callback to abstract `newSetOperationNode`. Then, you can pass an inline lambda which does the appropriate `new XXXNode` as the callback. Something like:

```java
public static SetOperationNode flatten(SetOperationNode node, Context context, NodeInstantiator instantiator)
{
    if (flattening not necessary) {
        return null;
    }   
    ...
    return instantiator.create(....);
}
```

(`NodeInstantiator` would be a single-method interface that takes the arguments needed to instantiate an `IntersectNode`, `UnionNode`, etc)

and in the rule:

```java
SetOperationNode result = flatten(node, IntersectNode::new);
if (result != null) {
    return Optional.ofPlanNode(result);
}
```

Reply:
That makes sense. I'll update accordingly.